src/solver/conv_asm_implicit_gemm_gtc_bwd.cpp

```diff
@@ -0,0 +1,318 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2020 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#include "miopen/solver.hpp"
+#include "miopen/handle.hpp"
+#include <miopen/conv/invokers/impl_gemm_dynamic.hpp>
+#include <miopen/generic_search.hpp>
+#include <miopen/gcn_asm_utils.hpp>
+#include "implicitgemm_util.hpp"
+
+namespace miopen {
+namespace solver {
+
+static inline const std::vector<TunableImplicitGemmGTCDynamic_t>&
+GetImplicitGemmGtcDynamicBwdTunablesList()
+{
+    // clang-format off
+    static const std::vector<TunableImplicitGemmGTCDynamic_t> tunables = {
+        { "bwd", "fp32",  16,   0, 128, 128,  16,  32,  32,   1,   1,   2,   2,   {2,   1,   1,   4},   {1,   8,   1,  32},   {2,   1,   1,   4},   {1,   8,   1,  32},   0},
```

Comment:
What is the semantics of things like "bwd" and "fp32" and how these are used in the code?

Reply:
Hi Artem,
'bwd' means one of the convolution directions, other two are 'fwd' and 'wrw', which you must already see in pr441 pr471. 
'fp32' means format of the data, which will be used in convolution.
both of them will be used to form the kernel name to be choosed, which will distinguish other kernels.

Reply:
Repeating of the same values looks really bad. Please remove these from the table and replace
```
kernel_name = pConfig->GetKernelName();
```
by
```
kernel_name = pConfig->GetKernelName("bwd", "fp32");
```
in `FindImplicitGemmGtcDynamicBwdKernel()`.

Reply:
Please also move definition of `TunableImplicitGemmGTCDynamic_t` from solver.hpp. This is internal detail of implementation of a family of asm igemm gtc solvers (perhaps you'll need to create a dedicated header to store such details).

Reply:
hi Artem,
As you know, for asm igemm, all kernels are physically generated by iGEMMgen[https://github.com/ROCmSoftwarePlatform/iGEMMgen/](url).
So actually TunableImplicitGemmGTCDynamic_t is the key shared structure, iGEMMgen use it to generate thousands of  asm kernels, and specific data member, like 'bwd', 'fwd', 'wrw','fp32' are used to generated the different kernels for different support senarior, it's like a config data base. 
For different convolution direction, there'll be different TunableImplicitGemmGTCDynamic_t list, you can refer to file: src/solver/conv_asm_implicit_gemm_gtc_bwd.cpp, src/solver/conv_asm_implicit_gemm_gtc_fwd.cpp, src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp.
So to make a conclusion, keep 'bwd', 'fwd', 'wrw' in the structure, will make solver more consistent with iGEMMgen, and make solver's config list more clear to specific convolution direction and data type. What do you think?

Reply:
> For different convolution direction, there'll be different TunableImplicitGemmGTCDynamic_t list

Of course I know that. And that is why repeating stuff like `"bwd", "fp32"` many times looks useless and confuses the reader. However, I am not going to insist on removing these.

BTW, is it so that the contents of "tunables" is generated from .config files (like this one: https://github.com/ROCmSoftwarePlatform/iGEMMgen/blob/master/config/igemm_bwd_gtc_gfx908.config) by iGEMMgen?

Reply:
Regardless of the above, moving the definition of `TunableImplicitGemmGTCDynamic_t` from solver.hpp to some header (common for asm igemm gtc solvers) should be done.